java-frontend/src/test/java/org/sonar/java/se/JavaCheckVerifier.java

```diff
@@ -252,50 +257,41 @@ private void validateAnalyzerMessageAttributes(Expectations.Issue attrs, Analyze
+    softAssertAttributeMatch(textSpan.endLine, attrs, END_LINE, flowAssertions);
+    softAssertAttributeMatch(normalizeColumn(textSpan.endCharacter), attrs, END_COLUMN, flowAssertions);
   }
 
-  private void validateFlows(List<List<AnalyzerMessage>> actual, List<String> expectedFlowIds) {
+  private void validateFlows(List<List<AnalyzerMessage>> actual, Set<String> expectedFlowIds) {
     Map<String, List<AnalyzerMessage>> foundFlows = new HashMap<>();
     List<List<AnalyzerMessage>> unexpectedFlows = new ArrayList<>();
     actual.forEach(f -> validateFlow(f, foundFlows, unexpectedFlows));
-    expectedFlowIds.removeAll(foundFlows.keySet());
-
-    assertExpectedAndMissingFlows(expectedFlowIds, unexpectedFlows);
+    Sets.SetView<String> missingFlows = Sets.difference(expectedFlowIds, foundFlows.keySet());
+    assertExpectedAndMissingFlows(missingFlows, unexpectedFlows);
     validateFoundFlows(foundFlows);
   }
 
-  private void assertExpectedAndMissingFlows(List<String> expectedFlowIds, List<List<AnalyzerMessage>> unexpectedFlows) {
-    if (expectedFlowIds.size() == 1 && expectedFlowIds.size() == unexpectedFlows.size()) {
-      assertSoleFlowDiscrepancy(expectedFlowIds.get(0), unexpectedFlows.get(0));
+  private void assertExpectedAndMissingFlows(Set<String> missingFlows, List<List<AnalyzerMessage>> unexpectedFlows) {
+    if (missingFlows.size() == 1 && missingFlows.size() == unexpectedFlows.size()) {
+      assertSoleFlowDiscrepancy(missingFlows.iterator().next(), unexpectedFlows.get(0));
     }
 
-    String unexpectedMsg = unexpectedFlows.stream()
-      .map(JavaCheckVerifier::flowToString)
-      .collect(joining("\n"));
-
-    String missingMsg = expectedFlowIds.stream().map(fid -> String.format("%s [%s]", fid, expectations.flowToLines(fid))).collect(joining(","));
+    flowAssertions.assertThat(unexpectedFlows).as("There should be no unexpected flows").isEmpty();
 
-    if (!unexpectedMsg.isEmpty() || !missingMsg.isEmpty()) {
-      unexpectedMsg = unexpectedMsg.isEmpty() ? "" : String.format("Unexpected flows: %s. ", unexpectedMsg);
-      missingMsg = missingMsg.isEmpty() ? "" : String.format("Missing flows: %s.", missingMsg);
-      Fail.fail(unexpectedMsg + missingMsg);
-    }
+    flowAssertions.assertThat(missingFlows).as("There should be no missing flows").isEmpty();
   }
 
   private void assertSoleFlowDiscrepancy(String expectedId, List<AnalyzerMessage> actualFlow) {
     Collection<Expectations.FlowComment> expected = expectations.flows.get(expectedId);
     List<Integer> expectedLines = expected.stream().map(f -> f.line).collect(Collectors.toList());
     List<Integer> actualLines = actualFlow.stream().map(AnalyzerMessage::getLine).collect(Collectors.toList());
-    assertThat(actualLines).as("Flow " + expectedId + " has line differences").isEqualTo(expectedLines);
+    flowAssertions.assertThat(actualLines).as("Flow " + expectedId + " has line differences").isEqualTo(expectedLines);
   }
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Add a call to 'assertAll' after all 'assertThat'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2970)